### PR TITLE
Add static assertion that uint32_t is unsigned int or wider

### PR DESCRIPTION
**Pull request by real-or-random**
Solves one item in #792 .

### CODE REVIEW

File: src/assumptions.h [hunk 1]
```diff
@@ -19,7 +21,11 @@ struct secp256k1_assumption_checker {
        allowed. */
     int dummy_array[(
         /* Bytes are 8 bits. */
-        CHAR_BIT == 8 &&
+        (CHAR_BIT == 8) &&
+
+        /* No integer promotion for uint32_t. This ensures that we can multiply uintXX_t values where XX >= 32
+           without signed overflow, which would be undefined behaviour. */
+        (UINT_MAX <= UINT32_MAX) &&
```
Comment (roconnor-blockstream) [inline src/assumptions.h:28]: I think `INT_MAX < UINT32_MAX` would more directly capture your intention here; however given the other rules of C, I do admit that what you have written ends up being equivalent.

Comment (real-or-random) [reply src/assumptions.h:28]: Yeah, I think our intuition is just different. My thought was that if `uint32_t` is at least as large as `unsigned int`, then no promotion can happen at all, neither to `signed int` nor to `unsigned int`.

Comment (roconnor-blockstream) [reply src/assumptions.h:28]: I think of the rule as if every value of a type can fit within a `signed int`, then it is subject to integer promotion.

Comment (sipa) [reply src/assumptions.h:28]: Both seem reasonable and harmless to me.

Comment (real-or-random) [reply src/assumptions.h:28]: Wanna ACK then?

**Review by elichai — APPROVED** [review]
Looks good.
ACK c0041b5cfca5efb160aa9a5616350069c89a8c29

**Comment by sipa** [comment]
utACK c0041b5cfca5efb160aa9a5616350069c89a8c29

**Pull request merged**